lib/handler/cache-handler.js

```diff
@@ -0,0 +1,202 @@
+'use strict'
+
+const util = require('../core/util.js')
+const DecoratorHandler = require('../handler/decorator-handler')
+const {
+  parseCacheControlHeader,
+  shouldRequestBeCached,
+  parseVaryHeader
+} = require('../util/cache.js')
+
+/**
+ * Takes a request, copies its data to a cache store, and passes it to the next
+ *  handler.
+ */
+class CacheHandler extends DecoratorHandler {
+  #globalOpts
+  #req
+  #handler
+  /**
+   * @type {import('../../types/cache-interceptor.d.ts').default.CacheStoreValue | undefined}
+   */
+  #value = undefined
+
+  /**
+   * @param {import('../../types/cache-interceptor.d.ts').default.CacheOptions} globalOpts
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @param {import('../../types/dispatcher.d.ts').default.DispatchHandlers} handler
+   */
+  constructor (globalOpts, req, handler) {
+    super(handler)
+
+    this.#globalOpts = globalOpts
+    this.#req = req
+    this.#handler = handler
+  }
+
+  onHeaders (
+    statusCode,
+    rawHeaders,
+    resume,
+    statusMessage,
+    headers = util.parseHeaders(rawHeaders)
+  ) {
+    const cacheControlHeader = headers['cache-control']
+
+    // Cache control header is missing or the status code isn't what we want,
+    //  let's not try to cache this response
+    if (!cacheControlHeader || !(statusCode in this.#globalOpts.statusCodes)) {
```

Comment:
```suggestion
    if (!cacheControlHeader || ![200,307].includes(statusCode)) {
```

Make supported status codes static for now